airflow/contrib/hooks/gcp_spanner_hook.py

```diff
@@ -0,0 +1,144 @@
+        :rtype: object
+        """
+        client = self.get_client(project_id)
+        instance = client.instance(instance_id)
+        if not instance.exists():
+            return None
+        instance.reload()
+        return instance
+
+    def create_instance(self, project_id, instance_id, configuration_name, node_count,
+                        display_name):
+        # type: (str, str, str, int, str) -> bool
+        return self.deploy_instance(project_id, instance_id, configuration_name,
+                                    node_count, display_name, lambda x: x.create())
+
+    def update_instance(self, project_id, instance_id, configuration_name, node_count,
+                        display_name):
+        # type: (str, str, str, int, str) -> bool
+        return self.deploy_instance(project_id, instance_id, configuration_name,
+                                    node_count, display_name, lambda x: x.update())
+
+    def deploy_instance(self, project_id, instance_id, configuration_name, node_count,
+                        display_name, func):
+        # type: (str, str, str, int, str, Callable[[Instance], Operation]) -> bool
+        """
+        Creates a new Cloud Spanner instance or, if an instance with the same name exists,
+        updates it.
+
+        :param project_id: The ID of the project which owns the instances, tables and
+        data.
+        :type project_id: str
+        :param instance_id: The ID of the instance.
+        :type instance_id: str
+        :param configuration_name: Name of the instance configuration defining how the
+            instance will be created. Required for instances which do not yet exist.
+        :type configuration_name: str
+        :param node_count: (Optional) Number of nodes allocated to the instance.
+        :type node_count: int
+        :param display_name: (Optional) The display name for the instance in the Cloud
+            Console UI. (Must be between 4 and 30 characters.) If this value is not set
+            in the constructor, will fall back to the instance ID.
+        :type display_name: str
+        :param func: Method of Instance to be called - either create() or update()
+        :type func: Callable
+        """
+        client = self.get_client(project_id)
+        instance = client.instance(instance_id,
+                                   configuration_name=configuration_name,
+                                   node_count=node_count,
+                                   display_name=display_name)
+        try:
+            operation = func(instance)  # type: Operation
+        except GoogleAPICallError as e:
+            self.log.error('An error occurred: {}. Exiting.'.format(e.message))
+            raise e
+
+        result = operation.result()
+        self.log.info(result)
+        return True
+
+    def delete_instance(self, project_id, instance_id):
```

Comment:
Can the delete_instance method utilize above "apply" method?

Reply:
Theoretically we could, I tested it, but to be honest I'm not fond of it. The `delete` is too dissimilar with `create` and `update`. They have different arguments and also both `create` and `update` return an `Operation`, whereas `delete` returns nothing.
So I'd rather stick with a separate `delete_instance()` method if that's okay with you.

Reply:
I have a feeling this would be a "leaky" abstraction. The `client.instance()` for the delete only takes in the `instance_id`, whereas for create and update it takes multiple parameters. Then there's the `Operation` returned / nothing returned case. So at least 2 `if`/`else` clauses, which to me would make the code less readable.

If you're okay with it I'd prefer to leave the `delete_instance()` method separate.